src/i2c/slave.rs

```diff
@@ -203,141 +215,181 @@ impl<'a, M: Mode> I2cSlave<'a, M> {
+
+                    if i2c.slvdat().read().data().bits() == ten_bit_info.second_byte {
+                        // We are selected, return a write transaction
+                        // The actual selection is only recorded once the user tries to read the
+                        // first byte from the master, as dropping before that should trigger a nack
+                        // anyway
+
+                        break Ok(Transaction::Write {
+                            handler: BlockingI2cSlaveWrite {
+                                info: self.base.info,
+                                ten_bit_info: &mut self.base.ten_bit_info,
+                            },
+                        });
+                    } else {
+                        // Not selected, mark us as such
+                        self.base.ten_bit_info = self.base.ten_bit_info.map(|mut v| {
+                            v.selected = false;
+                            v
+                        });
+
+                        continue;
                     }
                 } else {
-                    // Check slave is ready to receive
-                    if !i2c.stat().read().slvstate().is_slave_receive() {
-                        return Err(TransferError::ReadFail.into());
+                    // Read transaction, check if we are selected
+                    if ten_bit_info.selected {
+                        // We are selected, return a read transaction
+                        // acknowledge only when the user doesn't immediately drop but tries to
+                        // provide data to the master.
+
+                        break Ok(Transaction::Read {
+                            handler: BlockingI2cSlaveRead {
+                                info: self.base.info,
+                                should_ack_addr: true,
+                                _phantom: PhantomData,
+                            },
+                        });
+                    } else {
+                        // Not for us, nack
+                        i2c.slvctl().modify(|_, w| w.slvnack().nack());
+                        continue;
                     }
                 }
+            } else {
+                if addr & 1 == 0 {
+                    // Write transaction
+                    break Ok(Transaction::Write {
+                        handler: BlockingI2cSlaveWrite {
+                            info: self.base.info,
+                            ten_bit_info: &mut self.base.ten_bit_info,
+                        },
+                    });
+                } else {
+                    // Read transaction
+                    break Ok(Transaction::Read {
+                        handler: BlockingI2cSlaveRead {
+                            info: self.base.info,
+                            should_ack_addr: true,
+                            _phantom: PhantomData,
+                        },
+                    });
+                }
```

Comment:
I feel like this is too deeply nested, could we write this as a match, or split it out into separate functions (for 7-/10/bit addresses)?

Reply:
Done.